upgrade/v17xto180/upgrade.go

```diff
@@ -17,9 +21,179 @@ const (
+		return err
+	}
+
+	if err := upgradeBackupBackingImages(namespace, lhClient, resourceMaps); err != nil {
+		return err
+	}
+
+	if err := upgradeBackingImageDataSources(namespace, lhClient, resourceMaps); err != nil {
+		return err
+	}
+
 	return upgradeBackingImages(namespace, lhClient, resourceMaps)
 }
 
+func upgradeVolumes(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade volume failed")
+	}()
+
+	volumesMap, err := upgradeutil.ListAndUpdateVolumesInProvidedCache(namespace, lhClient, resourceMaps)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list all existing Longhorn volumes during the volume upgrade")
+	}
+
+	for _, v := range volumesMap {
+		backupTargetName := v.Spec.BackupTargetName
+		v.Spec.BackupTargetName = ensureBackupTargetSet(backupTargetName)
+	}
+
+	return nil
+}
+
+func ensureBackupTargetSet(backupTargetName string) string {
+	if backupTargetName == "" {
+		return types.DefaultBackupTargetName
+	}
+	return backupTargetName
+}
+
+func upgradeBackupVolumes(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade backup volume failed")
+	}()
+
+	backupVolumeMap, err := upgradeutil.ListAndUpdateBackupVolumesInProvidedCache(namespace, lhClient, resourceMaps)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list all existing Longhorn BackupVolumes during the Longhorn BackupVolume upgrade")
+	}
+
+	for _, bv := range backupVolumeMap {
+		backupTargetName := bv.Spec.BackupTargetName
+		bv.Spec.BackupTargetName = ensureBackupTargetSet(backupTargetName)
+
+		if bv.Spec.VolumeName == "" {
+			bv.Spec.VolumeName = bv.Name
+		}
+		bv.Labels = ensureLabel(bv.Labels, types.LonghornLabelBackupTarget, bv.Spec.BackupTargetName)
+		bv.Labels = ensureLabel(bv.Labels, types.LonghornLabelBackupVolume, bv.Spec.VolumeName)
+	}
+
+	return nil
+}
+
+func ensureLabel(labels map[string]string, key, value string) map[string]string {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	if _, exists := labels[key]; !exists {
+		labels[key] = value
+	}
+	return labels
+}
+
+func upgradeBackups(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade backup failed")
+	}()
+
+	backupMap, err := upgradeutil.ListAndUpdateBackupsInProvidedCache(namespace, lhClient, resourceMaps)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list all existing Longhorn Backups during the Longhorn Backup upgrade")
+	}
+
+	for _, b := range backupMap {
+		if b.Status.State != longhorn.BackupStateCompleted {
+			continue
+		}
```

Comment:
Why skip the backups that are not completed?

Reply:
Removed. Thanks.